Add support for Rockblock 9704 satellite modem

stephendade opened the pull request:
This PR adds support for the Rockblock 9704 satellite modem (https://www.groundcontrol.com/product/rockblock-9704/), using the MAVLink High Latency protocol. It sends a ``HIGH_LATENCY2`` packet once per 30 seconds.

The software required at the GCS to process messages is at https://github.com/stephendade/rockblock2mav/pull/5

I also:
- Added Lua bindings for CRC16 functions
- Disabled stream rate setting for MAVLink High Latency protocol

Tested in hardware with a CubeOrange+.

(Thanks to GroundControl for supplying the hardware and airtime)

stephendade commented:
> @stephendade not a good idea to mix in the streamrate changes with the 9704 support. Please split that out to a separate PR for separate discussion (I think I'm OK with it, but I'm wondering if we should add an option bit to allow people to use initial stream rates if they are still keen to do so).

Done. See #31331.

> My only hesitation is that I don't want to have a binding for every crc. How would you feel about a more generic binding where you can pass the polynomial?

My own thoughts are we only add bindings for commonly used CRC's. Looking through the AP_Scripting, the Xmodem CRC is used in >5 scripts, so that seems an obvious candidate.

tpwrules commented:
TL;DR steal the function from here, it's 3 times the speed: https://github.com/ArduPilot/ardupilot/blob/18e39549a5b3ea7b29c4a639a4f3202142058ddd/libraries/AP_Scripting/drivers/Hobbywing_DataLink.lua#L124 . I personally want to clean up the CRC code before adding bindings so that we can make a more generic one. I have also attached a 4x the speed version if you want.

It turns out the `DISABLE_INTERRUPTS_FOR_SCRIPT_RUN` doesn't really work, I am guessing that as soon as the memory allocator lock is released interrupts get turned back on. In any case, with some deep patches to fix that and the script at the end, I get the following results. They are very stable, varying maybe 1-2us per run (executed through my REPL):

```
CRC16 (bindings) is: 0x18B9                                                     
CRC16 (bindings) calculated in 14 us                                            
CRC16 (pure Lua) is: 0x18B9                                                     
CRC16 (pure Lua) calculated in 602 us                                           
CRC16 (opti Lua) is: 0x18B9                                                     
CRC16 (opti Lua) calculated in 488 us                                           
CRC16 (tabl Lua) is: 0x18B9                                                     
CRC16 (tabl Lua) calculated in 171 us                                           
luaC_fullgc calls: 1 
```

Other notes:
* 60us can be shaved by unpacking the whole string into a table and indexing that instead of calling `byte` for each byte. That is unexpectedly expensive.
* The binding can be made twice as fast by pointing it to `crc16_ccitt` as that also uses a table.

<details>

<summary> the code </summary>

```lua
-- Show various CRC calculations

--CRC16 Xmodem
inString = "Hello, world! This is a test string. 50 bytes long"

function crc_xmodem_update(crc, data)
    crc = crc ~ (data << 8)
    for i = 0, 7 do
        if (crc & 0x8000) ~= 0 then
            crc = (crc << 1) ~ 0x1021
        else
            crc = crc << 1
        end
    end
    return crc & 0xFFFF
end

function crc_xmodem_lua(data)
    local crc = 0
    for i = 1, #data do
        crc = crc_xmodem_update(crc, string.byte(data, i))
    end
    return crc
end

function crc_xmodem_lua_opti(s)
    local crc = 0
    local b = string.byte
    for i = 1, #s do
        local data = b(s, i)
        crc = crc ~ (data << 8)
        for i = 0, 7 do
            crc = crc << 1
            if (crc & 0x10000) ~= 0 then
                crc = crc ~ 0x1021
            end
        end
    end
    return crc & 0xFFFF
end

--[[
   xmodem CRC implementation thanks to https://github.com/cloudwu/skynet
   under MIT license
--]]
XMODEMCRC16Lookup = {
    [0] = 0x0000,
           0x1021,0x2042,0x3063,0x4084,0x50a5,0x60c6,0x70e7,
    0x8108,0x9129,0xa14a,0xb16b,0xc18c,0xd1ad,0xe1ce,0xf1ef,
    0x1231,0x0210,0x3273,0x2252,0x52b5,0x4294,0x72f7,0x62d6,
    0x9339,0x8318,0xb37b,0xa35a,0xd3bd,0xc39c,0xf3ff,0xe3de,
    0x2462,0x3443,0x0420,0x1401,0x64e6,0x74c7,0x44a4,0x5485,
    0xa56a,0xb54b,0x8528,0x9509,0xe5ee,0xf5cf,0xc5ac,0xd58d,
    0x3653,0x2672,0x1611,0x0630,0x76d7,0x66f6,0x5695,0x46b4,
    0xb75b,0xa77a,0x9719,0x8738,0xf7df,0xe7fe,0xd79d,0xc7bc,
    0x48c4,0x58e5,0x6886,0x78a7,0x0840,0x1861,0x2802,0x3823,
    0xc9cc,0xd9ed,0xe98e,0xf9af,0x8948,0x9969,0xa90a,0xb92b,
    0x5af5,0x4ad4,0x7ab7,0x6a96,0x1a71,0x0a50,0x3a33,0x2a12,
    0xdbfd,0xcbdc,0xfbbf,0xeb9e,0x9b79,0x8b58,0xbb3b,0xab1a,
    0x6ca6,0x7c87,0x4ce4,0x5cc5,0x2c22,0x3c03,0x0c60,0x1c41,
    0xedae,0xfd8f,0xcdec,0xddcd,0xad2a,0xbd0b,0x8d68,0x9d49,
    0x7e97,0x6eb6,0x5ed5,0x4ef4,0x3e13,0x2e32,0x1e51,0x0e70,
    0xff9f,0xefbe,0xdfdd,0xcffc,0xbf1b,0xaf3a,0x9f59,0x8f78,
    0x9188,0x81a9,0xb1ca,0xa1eb,0xd10c,0xc12d,0xf14e,0xe16f,
    0x1080,0x00a1,0x30c2,0x20e3,0x5004,0x4025,0x7046,0x6067,
    0x83b9,0x9398,0xa3fb,0xb3da,0xc33d,0xd31c,0xe37f,0xf35e,
    0x02b1,0x1290,0x22f3,0x32d2,0x4235,0x5214,0x6277,0x7256,
    0xb5ea,0xa5cb,0x95a8,0x8589,0xf56e,0xe54f,0xd52c,0xc50d,
    0x34e2,0x24c3,0x14a0,0x0481,0x7466,0x6447,0x5424,0x4405,
    0xa7db,0xb7fa,0x8799,0x97b8,0xe75f,0xf77e,0xc71d,0xd73c,
    0x26d3,0x36f2,0x0691,0x16b0,0x6657,0x7676,0x4615,0x5634,
    0xd94c,0xc96d,0xf90e,0xe92f,0x99c8,0x89e9,0xb98a,0xa9ab,
    0x5844,0x4865,0x7806,0x6827,0x18c0,0x08e1,0x3882,0x28a3,
    0xcb7d,0xdb5c,0xeb3f,0xfb1e,0x8bf9,0x9bd8,0xabbb,0xbb9a,
    0x4a75,0x5a54,0x6a37,0x7a16,0x0af1,0x1ad0,0x2ab3,0x3a92,
    0xfd2e,0xed0f,0xdd6c,0xcd4d,0xbdaa,0xad8b,0x9de8,0x8dc9,
    0x7c26,0x6c07,0x5c64,0x4c45,0x3ca2,0x2c83,0x1ce0,0x0cc1,
    0xef1f,0xff3e,0xcf5d,0xdf7c,0xaf9b,0xbfba,0x8fd9,0x9ff8,
    0x6e17,0x7e36,0x4e55,0x5e74,0x2e93,0x3eb2,0x0ed1,0x1ef0,
}

function crc_xmodem_lua_table(bytes)
    local t = XMODEMCRC16Lookup
    local b = string.byte
    local crc = 0
    for i=1,#bytes do
        crc = ((crc<<8) & 0xffff) ~ t[((crc>>8)~b(bytes, i))]
    end
    return crc
end

function check(fun)
    local micros = micros
    local inString = inString
    local crc

    local start = micros()
    local count = 3
    for i=1, count do
        crc = fun(inString)
    end
    local stop = micros()
    local delta = (stop - start)/count

    return crc, delta
end

function update()
    collectgarbage("stop")
    local smany = collectgarbage("howmany")
    collectgarbage("collect")

    local bcrc, bdelta = check(crc_xmodem)
    local lcrc, ldelta = check(crc_xmodem_lua)
    local ocrc, odelta = check(crc_xmodem_lua_opti)
    local tcrc, tdelta = check(crc_xmodem_lua_table)

    local emany = collectgarbage("howmany")

    collectgarbage("restart")

    print("CRC16 (bindings) is: " .. string.format("0x%04X", bcrc))
    print("CRC16 (bindings) calculated in " .. tostring(bdelta) .. " us")

    print("CRC16 (pure Lua) is: " .. string.format("0x%04X", lcrc))
    print("CRC16 (pure Lua) calculated in " .. tostring(ldelta) .. " us")

    print("CRC16 (opti Lua) is: " .. string.format("0x%04X", ocrc))
    print("CRC16 (opti Lua) calculated in " .. tostring(odelta) .. " us")

    print("CRC16 (tabl Lua) is: " .. string.format("0x%04X", tcrc))
    print("CRC16 (tabl Lua) calculated in " .. tostring(tdelta) .. " us")

    print("luaC_fullgc calls: "..tostring(emany-smany))
end
```
</details>

stephendade commented:
> I have also attached a 4x the speed version if you want.

Thanks! The ``crc_xmodem_lua_table`` works nicely for me. I've now incorporated that into the PR and removed the crc lua binding.

For reference, running your script on my CubeOrangePlus had quite different timings:
```
AP: CRC16 (bindings) is: 0x18B9
AP: CRC16 (bindings) calculated in 23 us
AP: CRC16 (opti Lua) is: 0x18B9
AP: CRC16 (opti Lua) calculated in 775 us
AP: CRC16 (tabl Lua) is: 0x18B9
AP: CRC16 (tabl Lua) calculated in 276 us
```

peterbarker commented:
OK, so this is a Lua-only change again.

@IamPete1 are you satisfied with this now?
